pkg/daemon/daemon.go

```diff
@@ -93,45 +93,61 @@ func (dn *Daemon) Run(stop <-chan struct{}) error {
 	return nil
 }
 
-// process starts the main loop that actually does all the work.
+// process is the main loop that actually does all the work. the flow goes
+// something like this -
+// 1. we restarted for some reason. the happy path reason we restarted is
+//    because of a machine reboot. validate the current machine state is the
+//    desired machine state. if we aren't try updating again. if we are, update
+//    the current state annotation accordingly.
+// 2. watch the desired config annotation, waiting for an update to be
+//    requested by the controller.
+// 3. if an update is requested by the controller, we assume that that means
+//    something changed and apply the desired config no matter what.
+// 4. the update function doesn't return right now, but at some point in the
+//    future if a reboot isn't required for an update it will. if it returns,
+//    validate the machine state and set the update to done.
+//
+// the only reason this function will return is if an error occurs. otherwise it
+// will keep trying to update the machine until it reboots.
 func (dn *Daemon) process() error {
-
-	// do a first pass before entering the main loop
-	if err := dn.syncOnce(); err != nil {
-		return err
-	}
-
 	for {
-		glog.V(2).Infof("Watching for node annotation updates on %q", dn.name)
-		if err := waitUntilUpdate(dn.kubeClient.CoreV1().Nodes(), dn.name); err != nil {
-			return err
-		}
-		if err := dn.syncOnce(); err != nil {
+		// validate machine state
+		isDesired, err := dn.isDesiredMachineState()
```

Comment:
We still need to enhance `isDesiredMachineState()` to check if the configs are reconcilable, right?

Reply:
I wasn't going to initially do that, but it would also fix the issue with this pr, and it's probably a good idea to do that, since it makes the validation function much more complete.